resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/MountedDependencyManagerImpl.scala

```diff
@@ -0,0 +1,284 @@
+
+  private def doUpload(files: Iterable[File], fileNamePrefix: String): String = {
+    val filesDir = Utils.createTempDir(namePrefix = fileNamePrefix)
+    val filesTgz = new File(filesDir, s"$fileNamePrefix.tgz")
+    Utils.tryWithResource(new FileOutputStream(filesTgz)) { filesOutputStream =>
+      CompressionUtils.writeTarGzipToStream(filesOutputStream, files.map(_.getAbsolutePath))
+    }
+    // TODO provide credentials properly when the staging server monitors the Kubernetes API.
+    val kubernetesCredentialsString = OBJECT_MAPPER.writer()
+      .writeValueAsString(KubernetesCredentials(None, None, None, None))
+    val labelsAsString = OBJECT_MAPPER.writer().writeValueAsString(podLabels)
+
+    val filesRequestBody = RequestBody.create(
+      okhttp3.MediaType.parse(MediaType.MULTIPART_FORM_DATA), filesTgz)
+
+    val kubernetesCredentialsBody = RequestBody.create(
+      okhttp3.MediaType.parse(MediaType.APPLICATION_JSON), kubernetesCredentialsString)
+
+    val namespaceRequestBody = RequestBody.create(
+      okhttp3.MediaType.parse(MediaType.TEXT_PLAIN), podNamespace)
+
+    val labelsRequestBody = RequestBody.create(
+      okhttp3.MediaType.parse(MediaType.APPLICATION_JSON), labelsAsString)
+
+    val service = retrofitClientFactory.createRetrofitClient(
+      stagingServerUri,
+      classOf[ResourceStagingServiceRetrofit],
+      stagingServiceSslOptions)
+    val uploadResponse = service.uploadResources(
+      labelsRequestBody, namespaceRequestBody, filesRequestBody, kubernetesCredentialsBody)
+    val rawSecret = getTypedResponseResult(uploadResponse)
+    rawSecret
+  }
+
+  override def configurePodToMountLocalDependencies(
+      driverContainerName: String,
+      initContainerSecret: Secret,
+      initContainerConfigMap: ConfigMap,
+      originalPodSpec: PodBuilder): PodBuilder = {
+    val sharedVolumeMounts = Seq[VolumeMount](
+      new VolumeMountBuilder()
+        .withName(DOWNLOAD_JARS_VOLUME_NAME)
+        .withMountPath(jarsDownloadPath)
+        .build(),
+      new VolumeMountBuilder()
+        .withName(DOWNLOAD_FILES_VOLUME_NAME)
+        .withMountPath(filesDownloadPath)
+        .build())
+
+    val initContainers = Seq(new ContainerBuilder()
+      .withName("driver-init")
+      .withImage(initContainerImage)
+      .withImagePullPolicy("IfNotPresent")
+      .addNewVolumeMount()
+        .withName(INIT_CONTAINER_PROPERTIES_FILE_VOLUME)
+        .withMountPath(INIT_CONTAINER_PROPERTIES_FILE_MOUNT_PATH)
+        .endVolumeMount()
+      .addNewVolumeMount()
+        .withName(INIT_CONTAINER_SECRETS_VOLUME_NAME)
+        .withMountPath(INIT_CONTAINER_SECRETS_VOLUME_MOUNT_PATH)
+        .endVolumeMount()
+      .addToVolumeMounts(sharedVolumeMounts: _*)
+      .addToArgs(INIT_CONTAINER_PROPERTIES_FILE_PATH)
+      .build())
+
+    // Make sure we don't override any user-provided init containers by just appending ours to
+    // the existing list.
+    val resolvedInitContainers = originalPodSpec
```

Comment:
Agreed, I like init containers, in the sense of being very idiomatic to kube, and also not requiring that logic to live on the user container images.  The flip side is that they seem a bit heavy-weight from the standpoint of having to instantiate the image, run it, etc.  Maybe @foxish has insights on this.